File: lua/distract/obstacles.lua
--- Solid platforms and hazards a plugin registered.
---
--- A provider is a function the editor calls on a debounced cadence — never per
--- tick per entity, because a Tree-sitter query per frame is a performance trap
--- — and its rectangles are pushed to whichever engine is running. Neither
--- engine discovers its own: only the editor can run a query, read a fold or see
--- where a function header is, and an engine that went looking would reintroduce
--- the divergence class the physics parity harness exists to catch.
---
--- In terminal cells. `external.lua` converts to overlay pixels at the boundary,
--- as it does for the floor. The three resolution rules mirror
--- `engine/src/obstacles.rs` function for function.

local M = {}

M.SOLID_PLATFORM = "solid_platform"
M.HAZARD = "hazard"

local KINDS = { [M.SOLID_PLATFORM] = true, [M.HAZARD] = true }

--- The most obstacles that are kept.
---
--- The physics pass is per entity per obstacle per frame, so a query over a
--- large file is bounded here rather than trusted. Matches
--- `obstacles::MAX_OBSTACLES`.
local MAX_OBSTACLES = 128

local providers = {}

--- The rectangles last collected, in cells.
local collected = {}

--- Whether the count cap has already been reported this session.
local has_warned_about_cap = false

--- Registers a provider.
---
--- @param provider function `function(win_id, buf_id) -> table[]`, each entry
---   `{ x, y, width, height, type }` in terminal cells
--- @return integer the provider's id, for `unregister_provider`
function M.register_provider(provider)
  if type(provider) ~= "function" then
    error("distract.register_obstacle_provider: provider must be a function")
  end
  table.insert(providers, provider)
  return #providers
end

function M.unregister_provider(id)
  if providers[id] == nil then
    return false
  end
  table.remove(providers, id)
  return true
end

function M.provider_count()
  return #providers
end

--- For tests, and for a full plugin reload.
function M.reset()
  providers = {}
  collected = {}
  has_warned_about_cap = false
end

--- Whether a rectangle is usable, and refuses it loudly when it is not.
---
--- A provider returning a malformed rectangle is a bug in that plugin. It is
--- skipped with one message rather than being allowed to reach the physics pass,
--- where a nil width becomes an arithmetic error at 30 frames a second.
local function is_valid(rect, source)
  if type(rect) ~= "table" then
    return false, "an obstacle must be a table"
  end
  for _, field in ipairs({ "x", "y", "width", "height" }) do
    if type(rect[field]) ~= "number" then
      return false, string.format("obstacle field '%s' must be a number", field)
    end
  end
  if rect.width <= 0 or rect.height <= 0 then
    return false, "an obstacle must have a positive width and height"
  end
  if not KINDS[rect.type] then
    return false,
      string.format(
        "obstacle type must be '%s' or '%s', got '%s'",
        M.SOLID_PLATFORM,
        M.HAZARD,
        tostring(rect.type)
      )
  end
  return true, source
end

--- Calls every provider and validates what comes back.
---
--- A provider that errors is reported once per collection and skipped; the rest
--- still contribute, because one broken plugin should not remove the ground
--- every other plugin registered.
---@param context { win: integer, buf: integer }|nil
---@return table[] the accepted rectangles, in cells
function M.collect(context)
  context = context or {}
  local win = context.win or vim.api.nvim_get_current_win()
  local buf = context.buf or vim.api.nvim_win_get_buf(win)

  local accepted = {}
  for index, provider in ipairs(providers) do
    local ok, result = pcall(provider, win, buf)
    if not ok then
      vim.notify(
        string.format("[Distract] Obstacle provider #%d failed: %s", index, tostring(result)),
        vim.log.levels.WARN
      )
    else
      for _, rect in ipairs(result or {}) do
        local valid, reason = is_valid(rect)
        if valid then
          table.insert(accepted, {
            x = rect.x,
            y = rect.y,
            width = rect.width,
            height = rect.height,
            type = rect.type,
          })
        else
          vim.notify(
            string.format("[Distract] Obstacle provider #%d: %s", index, reason),
            vim.log.levels.WARN
          )
        end
      end
    end
  end

  if #accepted > MAX_OBSTACLES then
    if not has_warned_about_cap then
      has_warned_about_cap = true
      vim.notify(
        string.format(
          "[Distract] %d obstacles registered; only the first %d are used. "
            .. "Narrow the provider's query.",
          #accepted,
          MAX_OBSTACLES
        ),
        vim.log.levels.WARN
      )
    end
    accepted = vim.list_slice(accepted, 1, MAX_OBSTACLES)
  end

  collected = accepted
  return collected
end

--- The rectangles last collected, in cells.
function M.rects()
  return collected
end

--- Replaces the collected list directly. For the engines' own tests.
function M.set_rects(rects)
  collected = rects or {}
end

local function spans(obstacle, left, right)
  return obstacle.x < right and left < obstacle.x + obstacle.width
end

--- The platform a falling entity has just crossed, or nil.
---
--- Crossing is what counts, not overlapping: an entity resting on a platform has
--- its feet exactly on the top edge and keeps re-crossing it every frame as
--- gravity re-accelerates it, which is what holds it there. Something moving
--- upward through a platform crosses nothing, which is what makes a jump onto one
--- work.
---
--- The highest crossed edge wins, so falling past several platforms in one frame
--- lands on the first one reached.
---@param footprint { left: number, top: number, width: number, height: number }
---@param feet_before number where the feet were before this frame's integration
---@return number|nil the platform's top edge
function M.crossed_platform(rects, footprint, feet_before)
  local feet_after = footprint.top + footprint.height
  if feet_after < feet_before then
    return nil
  end

  local highest = nil
  local right = footprint.left + footprint.width
  for _, obstacle in ipairs(rects) do
    if
      obstacle.type == M.SOLID_PLATFORM
      and spans(obstacle, footprint.left, right)
      and feet_before <= obstacle.y
      and obstacle.y <= feet_after
      and (highest == nil or obstacle.y < highest)
    then
      highest = obstacle.y
    end
  end
  return highest
end

--- The surface a grounded entity stands on, given the floor it would use.
---
--- A grounded state has no gravity — a walking cat's `y` never changes on its own
--- — so which surface it stands on is resolved rather than integrated. Only
--- platforms between the entity's own top and its floor count: one above its head
--- is scenery, one below the floor is unreachable.
---@return number the surface's top edge
function M.standing_surface(rects, footprint, floor)
  local highest = floor
  local right = footprint.left + footprint.width
  for _, obstacle in ipairs(rects) do
    if
      obstacle.type == M.SOLID_PLATFORM
      and spans(obstacle, footprint.left, right)
      and obstacle.y >= footprint.top
      and obstacle.y <= floor
      and obstacle.y < highest
    then
      highest = obstacle.y
    end
  end
  return highest
end

--- Where a hazard puts an entity that has walked into it.
---
--- The side is decided by which edge the entity is nearer to, so one that arrived
--- from the left is returned to the left. An entity exactly centred on a hazard is
--- returned the way its heading says it came.
---@return table|nil `{ x = number, heading_x = number }`
function M.deflection(rects, footprint, heading_x)
  local right = footprint.left + footprint.width
  local feet = footprint.top + footprint.height
  for _, obstacle in ipairs(rects) do
    if
      obstacle.type == M.HAZARD
      and spans(obstacle, footprint.left, right)
      and obstacle.y < feet
      and footprint.top < obstacle.y + obstacle.height
    then
      local overlap_from_left = right - obstacle.x
      local overlap_from_right = (obstacle.x + obstacle.width) - footprint.left
      local came_from_left
      if overlap_from_left == overlap_from_right then
        came_from_left = heading_x >= 0
      else
        came_from_left = overlap_from_left < overlap_from_right
      end

      if came_from_left then
        return { x = obstacle.x - footprint.width, heading_x = -1 }
      end
      return { x = obstacle.x + obstacle.width, heading_x = 1 }
    end
  end
  return nil
end

return M

File: engine/src/obstacles.rs
//! Solid ground and hazards a plugin registered, and the geometry that resolves
//! them.
//!
//! Collected in Neovim and pushed here, never discovered by this engine: only
//! the editor can run a Tree-sitter query or read a fold, and an engine that
//! went looking for its own obstacles is the divergence class the physics parity
//! harness exists to catch — exactly as with the floor.
//!
//! In overlay pixels, matching every other coordinate the engine is sent. The
//! resolution rules below are mirrored line for line by `lua/distract/obstacles.lua`.

use std::collections::HashMap;

use serde::{Deserialize, Serialize};

/// The most obstacles a client may push.
///
/// A Tree-sitter query over a large file can produce thousands of matches; the
/// physics pass is per entity per obstacle per frame, so the list is bounded at
/// the boundary rather than trusted.
pub const MAX_OBSTACLES: usize = 128;

/// What an obstacle does to an entity that reaches it.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ObstacleKind {
    /// A one-way floor: it catches an entity falling onto it from above and is
    /// what a grounded entity walks along. Passing upward through it is free,
    /// which is what makes a jump onto a platform work.
    #[serde(rename = "solid_platform", alias = "platform")]
    SolidPlatform,
    /// A vertical wall an entity turns away from.
    #[serde(rename = "hazard")]
    Hazard,
}

/// One registered rectangle, in overlay pixels.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct Obstacle {
    pub x: f32,
    pub y: f32,
    pub width: f32,
    pub height: f32,
    #[serde(rename = "type", alias = "kind")]
    pub kind: ObstacleKind,
}

impl Obstacle {
    pub fn right(&self) -> f32 {
        self.x + self.width
    }

    fn spans(&self, left: f32, right: f32) -> bool {
        self.x < right && left < self.right()
    }
}

/// Reads an obstacle list, accepting `{}` as well as `[]`.
///
/// `vim.json.encode` writes an empty Lua table as `{}`, and a session whose
/// provider currently finds nothing sends exactly that. Rejecting the encoding
/// would make "no platforms right now" an error on the overlay and a no-op in the
/// terminal. A *keyed* table is still an error: that is a mistake, not an empty
/// list.
///
/// # Errors
/// When the value is neither a list of obstacles nor an empty table.
pub fn list_allowing_empty_table<'de, D>(deserializer: D) -> Result<Vec<Obstacle>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    #[derive(Deserialize)]
    #[serde(untagged)]
    enum Obstacles {
        List(Vec<Obstacle>),
        Table(HashMap<String, serde::de::IgnoredAny>),
    }

    match Option::<Obstacles>::deserialize(deserializer)? {
        None => Ok(Vec::new()),
        Some(Obstacles::List(obstacles)) => Ok(obstacles),
        Some(Obstacles::Table(table)) if table.is_empty() => Ok(Vec::new()),
        Some(Obstacles::Table(_)) => Err(serde::de::Error::custom(
            "obstacles must be a list; a keyed table is a mistake, not an empty list",
        )),
    }
}

/// The footprint an entity occupies, in overlay pixels.
#[derive(Debug, Clone, Copy)]
pub struct Footprint {
    pub left: f32,
    pub top: f32,
    pub width: f32,
    pub height: f32,
}

impl Footprint {
    pub fn right(&self) -> f32 {
        self.left + self.width
    }

    pub fn feet(&self) -> f32 {
        self.top + self.height
    }
}

/// Which way an entity was pushed out of a hazard.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PushDirection {
    /// Turned back to the left, so its heading is now negative.
    Left,
    /// Turned back to the right.
    Right,
}

/// Where an entity is put when it is turned away from a hazard.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Deflection {
    pub x: f32,
    pub direction: PushDirection,
}

/// The platform a falling entity has just crossed, if any.
///
/// Crossing is what counts, not overlapping: an entity resting on a platform has
/// its feet exactly on the top edge and keeps re-crossing it every frame as
/// gravity re-accelerates it, which is what holds it there. Something moving
/// upward through a platform crosses nothing.
///
/// The highest crossed edge wins, so falling past several platforms in one frame
/// lands on the first one reached rather than the last one tested.
pub fn crossed_platform(
    obstacles: &[Obstacle],
    footprint: Footprint,
    feet_before: f32,
) -> Option<f32> {
    let feet_after = footprint.feet();
    if feet_after < feet_before {
        return None;
    }

    obstacles
        .iter()
        .filter(|obstacle| obstacle.kind == ObstacleKind::SolidPlatform)
        .filter(|obstacle| obstacle.spans(footprint.left, footprint.right()))
        .filter(|obstacle| feet_before <= obstacle.y && obstacle.y <= feet_after)
        .map(|obstacle| obstacle.y)
        .min_by(|left, right| left.total_cmp(right))
}

/// The surface a grounded entity stands on, given the floor it would otherwise
/// use.
///
/// A grounded state has no gravity — a walking cat's `y` never changes on its
/// own — so standing on a platform is a matter of which surface is under its
/// feet, resolved every frame. Only platforms between the entity's own top and
/// its floor count: one above its head is scenery, and one below the floor is
/// unreachable.
pub fn standing_surface(obstacles: &[Obstacle], footprint: Footprint, floor: f32) -> f32 {
    obstacles
        .iter()
        .filter(|obstacle| obstacle.kind == ObstacleKind::SolidPlatform)
        .filter(|obstacle| obstacle.spans(footprint.left, footprint.right()))
        .filter(|obstacle| obstacle.y >= footprint.top && obstacle.y <= floor)
        .map(|obstacle| obstacle.y)
        .min_by(|left, right| left.total_cmp(right))
        .unwrap_or(floor)
}

/// Where a hazard puts an entity that has walked into it.
///
/// The side is decided by which edge the entity is nearer to, so an entity that
/// arrived from the left is returned to the left. A hazard the entity is exactly
/// centred on returns it the way it came, which is the direction its heading
/// already describes.
pub fn deflection(
    obstacles: &[Obstacle],
    footprint: Footprint,
    heading_x: f32,
) -> Option<Deflection> {
    let hazard = obstacles
        .iter()
        .filter(|obstacle| obstacle.kind == ObstacleKind::Hazard)
        .find(|obstacle| {
            obstacle.spans(footprint.left, footprint.right())
                && obstacle.y < footprint.feet()
                && footprint.top < obstacle.y + obstacle.height
        })?;

    let overlap_from_left = footprint.right() - hazard.x;
    let overlap_from_right = hazard.right() - footprint.left;
    let came_from_left = if overlap_from_left == overlap_from_right {
        heading_x >= 0.0
    } else {
        overlap_from_left < overlap_from_right
    };

    if came_from_left {
        Some(Deflection {
            x: hazard.x - footprint.width,
            direction: PushDirection::Left,
        })
    } else {
        Some(Deflection {
            x: hazard.right(),
            direction: PushDirection::Right,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn platform(x: f32, y: f32, width: f32) -> Obstacle {
        Obstacle {
            x,
            y,
            width,
            height: 1.0,
            kind: ObstacleKind::SolidPlatform,
        }
    }

    fn hazard(x: f32, y: f32, width: f32, height: f32) -> Obstacle {
        Obstacle {
            x,
            y,
            width,
            height,
            kind: ObstacleKind::Hazard,
        }
    }

    fn footprint(left: f32, top: f32) -> Footprint {
        Footprint {
            left,
            top,
            width: 20.0,
            height: 10.0,
        }
    }

    #[test]
    fn an_empty_lua_table_is_an_empty_obstacle_list() {
        #[derive(Deserialize)]
        struct Wrapper {
            #[serde(default, deserialize_with = "list_allowing_empty_table")]
            obstacles: Vec<Obstacle>,
        }

        let empty: Wrapper = serde_json::from_str(r#"{"obstacles":{}}"#)
            .expect("vim.json.encode writes an empty table as {}");
        assert!(empty.obstacles.is_empty());

        let absent: Wrapper = serde_json::from_str("{}").expect("an absent list is empty");
        assert!(absent.obstacles.is_empty());

        let listed: Wrapper = serde_json::from_str(
            r#"{"obstacles":[{"x":1,"y":2,"width":3,"height":4,"type":"solid_platform"}]}"#,
        )
        .expect("a real list parses");
        assert_eq!(listed.obstacles.len(), 1);
        assert_eq!(listed.obstacles[0].kind, ObstacleKind::SolidPlatform);

        assert!(
            serde_json::from_str::<Wrapper>(r#"{"obstacles":{"first":{"x":1}}}"#).is_err(),
            "a keyed table is a mistake worth reporting"
        );
    }

    #[test]
    fn a_falling_entity_lands_on_the_platform_it_crossed() {
        let obstacles = [platform(0.0, 100.0, 200.0)];
        // Feet were at 95, are now at 105: the edge at 100 was crossed.
        let landed = crossed_platform(&obstacles, footprint(10.0, 95.0), 85.0);
        assert_eq!(landed, Some(100.0));
    }

    #[test]
    fn an_entity_resting_on_a_platform_keeps_being_caught_by_it() {
        let obstacles = [platform(0.0, 100.0, 200.0)];
        // Resting: feet exactly on the edge, then gravity nudges them below it.
        let landed = crossed_platform(&obstacles, footprint(10.0, 90.5), 100.0);
        assert_eq!(landed, Some(100.0));
    }

    #[test]
    fn a_platform_is_one_way() {
        let obstacles = [platform(0.0, 100.0, 200.0)];
        // Moving upward through the same edge: feet were below, are now above.
        assert_eq!(
            crossed_platform(&obstacles, footprint(10.0, 80.0), 105.0),
            None
        );
    }

    #[test]
    fn a_platform_beside_the_entity_catches_nothing() {
        let obstacles = [platform(500.0, 100.0, 200.0)];
        assert_eq!(
            crossed_platform(&obstacles, footprint(10.0, 95.0), 85.0),
            None
        );
    }

    #[test]
    fn falling_past_several_platforms_lands_on_the_first_one_reached() {
        let obstacles = [
            platform(0.0, 160.0, 200.0),
            platform(0.0, 120.0, 200.0),
            platform(0.0, 200.0, 200.0),
        ];
        let landed = crossed_platform(&obstacles, footprint(10.0, 250.0), 100.0);
        assert_eq!(landed, Some(120.0));
    }

    #[test]
    fn a_grounded_entity_stands_on_the_highest_platform_under_its_feet() {
        let obstacles = [platform(0.0, 300.0, 400.0), platform(0.0, 250.0, 400.0)];
        let surface = standing_surface(&obstacles, footprint(10.0, 240.0), 400.0);
        assert_eq!(surface, 250.0);
    }

    #[test]
    fn a_grounded_entity_falls_back_to_the_floor_when_it_walks_off_the_end() {
        let obstacles = [platform(0.0, 250.0, 100.0)];
        let surface = standing_surface(&obstacles, footprint(300.0, 240.0), 400.0);
        assert_eq!(surface, 400.0);
    }

    #[test]
    fn a_platform_above_the_entitys_head_is_scenery() {
        let obstacles = [platform(0.0, 50.0, 400.0)];
        let surface = standing_surface(&obstacles, footprint(10.0, 240.0), 400.0);
        assert_eq!(surface, 400.0);
    }

    #[test]
    fn a_hazard_returns_an_entity_the_way_it_came() {
        let obstacles = [hazard(100.0, 0.0, 20.0, 400.0)];
        // Arrived from the left: its right edge is just inside the hazard.
        let deflected = deflection(&obstacles, footprint(95.0, 100.0), 1.0)
            .expect("the footprint overlaps the hazard");
        assert_eq!(deflected.direction, PushDirection::Left);
        assert_eq!(deflected.x, 80.0);

        // Arrived from the right.
        let deflected = deflection(&obstacles, footprint(105.0, 100.0), -1.0)
            .expect("the footprint overlaps the hazard");
        assert_eq!(deflected.direction, PushDirection::Right);
        assert_eq!(deflected.x, 120.0);
    }

    #[test]
    fn a_hazard_the_entity_is_not_touching_deflects_nothing() {
        let obstacles = [hazard(100.0, 0.0, 20.0, 40.0)];
        // Horizontally over it, vertically well below.
        assert_eq!(deflection(&obstacles, footprint(95.0, 300.0), 1.0), None);
        // Vertically level, horizontally clear.
        assert_eq!(deflection(&obstacles, footprint(400.0, 10.0), 1.0), None);
    }

    #[test]
    fn a_platform_never_deflects_and_a_hazard_never_supports() {
        let obstacles = [
            platform(100.0, 100.0, 20.0),
            hazard(100.0, 100.0, 20.0, 20.0),
        ];
        assert_eq!(
            deflection(&obstacles[..1], footprint(95.0, 95.0), 1.0),
            None
        );
        assert_eq!(
            standing_surface(&obstacles[1..], footprint(95.0, 50.0), 400.0),
            400.0
        );
    }
}
